crates/ecs/src/lib.rs

```diff
@@ -585,27 +871,51 @@ impl<'a, Component> Deref for Read<'a, Component> {
     }
 }
 
-impl<Component: Send + Sync + 'static + Sized> SystemParameter for Read<'_, Component> {
-    type BorrowedData<'components> = ReadComponentVec<'components, Component>;
+#[inline(always)]
+fn get_and_inc(value: &mut usize) -> usize {
+    let tmp = *value;
+    *value += 1;
+    tmp
+}
 
-    fn borrow(world: &World) -> Self::BorrowedData<'_> {
-        world.borrow_component_vec::<Component>()
+impl<Component: Debug + Send + Sync + 'static + Sized> SystemParameter for Read<'_, Component> {
+    type BorrowedData<'components> = (
+        usize,
+        Vec<(usize, ReadComponentVec<'components, Component>)>,
+    );
+
+    fn borrow<'a>(world: &'a World, signature: &[TypeId]) -> Self::BorrowedData<'a> {
+        (
+            0,
+            world
+                .borrow_component_vecs_with_signature::<Component>(signature)
+                .into_iter()
+                .map(|component_vec| (0, component_vec))
+                .collect(),
+        )
     }
 
-    unsafe fn fetch_parameter(
-        borrowed: &mut Self::BorrowedData<'_>,
-        entity: Entity,
-    ) -> Option<Self> {
-        if let Some(component_vec) = borrowed {
-            if let Some(Some(component)) = component_vec.get(entity.id) {
-                return Some(Self {
-                    // The caller is responsible to only use the
-                    // returned value when BorrowedData is still in scope.
-                    #[allow(trivial_casts)]
-                    output: &*(component as *const Component),
-                });
+    unsafe fn fetch_parameter(borrowed: &mut Self::BorrowedData<'_>) -> Option<Option<Self>> {
+        if let Some(archetype) = (borrowed.1).get_mut(borrowed.0) {
```

Comment:
when working with tuples it's usually best to destructure them ASAP, to avoid obtuse `.0`s and `.1`s everywhere.

I.e.
```rust
let (archetype_index, component_vec) = borrowed;
```
and then you use `archetype_index` and `component_vec` instead.